pkg/bgpv1/manager/pod_ip_pool_reconciler.go

```diff
@@ -0,0 +1,232 @@
+	"context"
+	"fmt"
+	"net/netip"
+
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	v2api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	v2alpha1api "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+)
+
+type PodIPPoolReconcilerOut struct {
+	cell.Out
+
+	Reconciler ConfigReconciler `group:"bgp-config-reconciler"`
+}
+
+type PodIPPoolReconciler struct {
+	poolStore BGPCPResourceStore[*v2alpha1api.CiliumPodIPPool]
+}
+
+func NewPodIPPoolReconciler(poolStore BGPCPResourceStore[*v2alpha1api.CiliumPodIPPool]) PodIPPoolReconcilerOut {
+	if poolStore == nil {
+		return PodIPPoolReconcilerOut{}
+	}
+
+	return PodIPPoolReconcilerOut{
+		Reconciler: &PodIPPoolReconciler{
+			poolStore: poolStore,
+		},
+	}
+}
+
+func (r *PodIPPoolReconciler) Priority() int {
+	return 50
+}
+
+func (r *PodIPPoolReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
+	lp := r.populateLocalPools(p.CiliumNode)
+
+	if err := r.fullReconciliation(ctx, p.CurrentServer, p.DesiredConfig, lp); err != nil {
+		return fmt.Errorf("full reconciliation failed: %w", err)
+	}
+
+	return nil
+}
+
+// populateLocalPools returns a map of allocated multi-pool IPAM CIDRs of the local CiliumNode,
+// keyed by the pool name.
+func (r *PodIPPoolReconciler) populateLocalPools(localNode *v2api.CiliumNode) map[string][]netip.Prefix {
+	var (
+		l = log.WithFields(
+			logrus.Fields{
+				"component": "manager.podIPPoolReconciler",
+			},
+		)
+	)
+
+	if localNode == nil {
+		return nil
+	}
+
+	lp := make(map[string][]netip.Prefix)
+	for _, pool := range localNode.Spec.IPAM.Pools.Allocated {
+		var prefixes []netip.Prefix
+		for _, cidr := range pool.CIDRs {
+			if p, err := cidr.ToPrefix(); err == nil {
+				prefixes = append(prefixes, *p)
+			} else {
+				l.Errorf("invalid ipam pool cidr %v: %v", cidr, err)
+			}
+		}
+		lp[pool.Pool] = prefixes
+	}
+
+	return lp
+}
+
+// fullReconciliation reconciles all pod ip pools.
+func (r *PodIPPoolReconciler) fullReconciliation(ctx context.Context,
+	sc *ServerWithConfig,
+	newc *v2alpha1api.CiliumBGPVirtualRouter,
+	localPools map[string][]netip.Prefix) error {
+	// Loop over all existing announcements, delete announcements for pod ip pools that no longer exist.
+	for poolKey := range sc.PodIPPoolAnnouncements {
+		_, found, err := r.poolStore.GetByKey(poolKey)
+		if err != nil {
+			return fmt.Errorf("failed to get pod ip pool from resource store: %w", err)
+		}
+		// If the pod ip pool no longer exists, withdraw all associated routes.
+		if !found {
+			if err := r.withdrawPool(ctx, sc, poolKey); err != nil {
+				return fmt.Errorf("failed to withdraw pod ip pool: %w", err)
+			}
+			continue
+		}
+	}
+
+	// Loop over all pod ip pools, reconcile any updates to the pool.
+	iter := r.poolStore.IterKeys()
```

Comment:
Nit: you can also use `r.poolStore.List()` instead of the iterator

Reply:
@rastislavs thanks for the guidance. Commit 572b94b uses the returned value from `r.poolStore.List()` as the iterator.